arbos/constraints/multi_dimensional.go

```diff
@@ -0,0 +1,213 @@
+)
+
+// Fixed flat layout for a Multi-Constraint:
+// [0] target (uint64)
+// [1] adjustmentWindow (uint64)
+// [2] backlog (uint64)
+// [3] sumWeights (uint64)
+// [4..4+NumResourceKind-1] weighted resources (uint64 each)
+
+const (
+	targetOffset uint64 = iota
+	adjustmentWindowOffset
+	backlogOffset
+	sumWeightsOffset
+	weightedResourcesBaseOffset
+)
+
+// MultiGasConstraint defines a pricing constraint that combines several
+// gas resource types, each with a corresponding weight (0 = unused).
+type MultiGasConstraint struct {
+	target            storage.StorageBackedUint64
+	adjustmentWindow  storage.StorageBackedUint64
+	backlog           storage.StorageBackedUint64
+	sumWeights        storage.StorageBackedUint64
+	weightedResources [multigas.NumResourceKind]storage.StorageBackedUint64
+}
+
+// OpenMultiGasConstraint opens or initializes a constraint in the given storage subspace.
+func OpenMultiGasConstraint(sto *storage.Storage) *MultiGasConstraint {
+	c := &MultiGasConstraint{
+		target:           sto.OpenStorageBackedUint64(targetOffset),
+		adjustmentWindow: sto.OpenStorageBackedUint64(adjustmentWindowOffset),
+		backlog:          sto.OpenStorageBackedUint64(backlogOffset),
+		sumWeights:       sto.OpenStorageBackedUint64(sumWeightsOffset),
+	}
+	for i := range int(multigas.NumResourceKind) {
+		// #nosec G115 safe: NumResourceKind < 2^32
+		offset := weightedResourcesBaseOffset + uint64(i)
+		c.weightedResources[i] = sto.OpenStorageBackedUint64(offset)
+	}
+	return c
+}
+
+// Clear resets the constraint and all weighted resources.
+func (c *MultiGasConstraint) Clear() error {
+	if err := c.target.Clear(); err != nil {
+		return err
+	}
+	if err := c.adjustmentWindow.Clear(); err != nil {
+		return err
+	}
+	if err := c.backlog.Clear(); err != nil {
+		return err
+	}
+	if err := c.sumWeights.Clear(); err != nil {
+		return err
+	}
+	for i := range int(multigas.NumResourceKind) {
+		if err := c.weightedResources[i].Clear(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// SetResourceWeights assigns per-resource weight multipliers for this constraint.
+func (c *MultiGasConstraint) SetResourceWeights(weights map[uint8]uint64) error {
+	var total uint64
+	for kind, weight := range weights {
+		if _, err := multigas.CheckResourceKind(kind); err != nil {
+			return err
+		}
+		total = arbmath.SaturatingUAdd(total, weight)
+	}
+	for i := range int(multigas.NumResourceKind) {
+		// #nosec G115 safe: NumResourceKind < 2^32
+		weight, ok := weights[uint8(i)]
+		if ok {
+			if err := c.weightedResources[i].Set(weight); err != nil {
+				return err
+			}
+		} else {
+			// zeroise weight for unspecified resources
+			if err := c.weightedResources[i].Set(0); err != nil {
+				return err
+			}
+		}
+	}
+	return c.sumWeights.Set(total)
+}
+
+// ConstraintContribution returns the normalized backlog contribution for a specific resource kind.
+func (c *MultiGasConstraint) ConstraintContribution(kind uint8) (uint64, error) {
```

Comment:
```suggestion
// ComputeExponent returns the exponent of the given constraint for the given resource kind.
func (c *MultiGasConstraint) ComputeExponent(kind uint8) (uint64, error) {
```
This function name is misleading because it is actually returning the constraint exponent for the given resource.

Reply:
fixed